### ISSUE TITLE
Netty will not complain if port 8080 is already used

### ISSUE THREAD

**Issue by delabassee** [oracle/helidon#249]
## Environment Details
* Helidon Version: SE latest (via `mvn archetype`)
* Helidon SE or Helidon MP : SE
* JDK version: 1.8.0_171
* OS: OS X 10.13.6 (17G65)
* Docker version (if applicable): N/A
----------

## Problem Description

When the 8080 port is already used by another process, Netty will sill claim that it is up and running!

## Steps to reproduce

- make sure that you have another process using port 8080

- `mvn archetype:generate...`

-> test will fail ` MainTest.testHelloWorld:60 HTTP response1 ==> expected: <200> but was: <404>`

or

- add `<skipTests>true</skipTests>` in the pom.xml

- `mvn package`

- `java -jar target/quickstart-se.jar`

```
[DEBUG] (main) Using Console logging
2018.12.10 10:46:22 INFO io.netty.util.internal.PlatformDependent Thread[main,5,main]: Your platform does not provide complete low-level API for accessing direct buffers reliably. Unless explicitly requested, heap buffer will always be preferred to avoid potential system instability.
2018.12.10 10:46:22 INFO io.helidon.webserver.netty.NettyWebServer Thread[nioEventLoopGroup-2-1,10,main]: Channel '@default' started: [id: 0x8e333db7, L:/0:0:0:0:0:0:0:0:8080]
WEB server is up! http://localhost:8080
```

But clearly, WEB server is not up! I would accept an error message telling me that it wasn't able to start as the port was already occupied.

This issue also apply to Helidon MP:

```
2018.12.10 11:18:21 INFO io.helidon.webserver.netty.NettyWebServer Thread[nioEventLoopGroup-2-1,10,main]: Channel '@default' started: [id: 0x3d2cbe90, L:/0:0:0:0:0:0:0:0:8080]
2018.12.10 11:18:21 INFO io.helidon.microprofile.server.ServerImpl Thread[nioEventLoopGroup-2-1,10,main]: Server started on http://localhost:8080 (and all other host addresses) in 97 milliseconds.
http://localhost:8080/greet
```

**Comment by tomas-langer** [comment]
I have experienced this on MacOS only. Please do netstat -an | grep 8080 and post the result.
Netty actually does listen - there is a problem between IPv4 and IPv6 sockets.

**Comment by delabassee** [comment]
I'm using 0.10.5.

Netty **does not** listen on port 8080. In my case, it was Fn.

```
netstat -an | grep 8080 
tcp6       0      0  ::1.8080               *.*                    LISTEN
tcp4       0      0  *.8080                 *.*                    LISTEN

curl localhost:8080
{"goto":"https://github.com/fnproject/fn","hello":"world!"}
```

**Comment by tomas-langer** [comment]
Hi, according to the log it seems that netty started on ipv6 (L:/0:0:0:0:0:0:0:0:8080)
also your netstat output shows a process listening on tcp6 - maybe that is netty?

I am just trying to pinpoint the problem and possible solution. 
Maybe by default, we should try to occupy both v4 and v6 and fail if either is already used. The current behavior seems to succeed if either succeeds (even if the other fails).

Would that be a good approach?

We should have a configuration option that would allow us to choose whether to use tcp4 or tcp6 or both.

**Comment by romain-grecourt** [comment]
@tomas-langer @spericas @barchetta How do you feel about targeting this one for 1.0 ?

**Comment by spericas** [comment]
@romain-grecourt Yes, we should fix this for 1.0.

**Comment by tomas-langer** [comment]
I have trouble reproducing this.
I can run helidon on the same port twice only when using different addresses (e.g. one on 0.0.0.0 and another on my local address):
```
# this is a webserver started with InetAddress.getLocalHost()
tcp4       0      0  192.168.0.104.8080     *.*                    LISTEN
# this is a webserver started with address 0.0.0.0
tcp46      0      0  *.8080                 *.*                    LISTEN  
```
Which is a different case from the one described by @delabassee ; when I tried to start the server on a port already occupied, I always get a 
`INFO: Channel '@default' startup failed with message 'Address already in use'.`

**Comment by tomas-langer** [comment]
I have also tried using a port that is already occupied on tcp4 and tcp6:
tcp4       0      0  *.5900                 *.*                    LISTEN
tcp6       0      0  *.5900                 *.*                    LISTEN
Still get a nice `Address already in use` exception :(

**Comment by tomas-langer** [comment]
This problem manifests only when listening on the all hosts ("0.0.0.0" or "::"), as java automatically binds these address to both IP v4 and IP v6 - unfortunately there is no option to modify this behavior.
So in our case the docker seems to start on tcp4 and tcp6 sockets which are separated and java starts on a combined tcp46 socket, which for some reason do not conflict on MacOS.
I could not find a way to invoke the tcp46 socket from curl or browser, so it is in fact a "zombie" server with no possibility to access it, even though it started correctly.
When I start webserver on the loopback address (`127.0.0.1`), it can be reached through `http://127.0.0.1`, though `http://localhost` hits the docker machine.
I need to request help from somebody with better understanding of the network stack to find if there is a fix to this.

**Comment by tomas-langer** [comment]
Information about java binding to ports:
https://docs.oracle.com/javase/8/docs/technotes/guides/net/ipv6_guide/index.html

**Comment by spericas** [comment]
I'll take a look ...

**Comment by spericas** [comment]
I was able to reproduce the issue following @tomas-langer instructions. Here is the output of netstat when the 2 processes are running (just like it was for @tomas-langer):

```
(~) >> netstat -a | grep 9411
# Helidon
tcp46      0      0  *.9411                 *.*                    LISTEN     
# Docker
tcp6       0      0  localhost.9411         *.*                    LISTEN     
tcp4       0      0  *.9411                 *.*                    LISTEN 
```

What's interesting about this output is that Docker is **_only_** binding to `localhost` on ipv6, while it is binding on **_all_** interfaces on ipv4. 

When the Helidon (Java) app starts it tries to bind on all protocols (tcp4 or tcp6) and all interfaces and finds that tcp6 is available in for all interfaces except `localhost`, and it thus succeeds. Here is the confirmation of this theory:

```
curl "http://localhost:9411/greet" -> Docker
curl "http://127.0.0.1:9411/greet" -> Docker (the lo0 interface)
curl "http://10.0.0.201:9411/greet" -> Docker (the en0 interface)
curl "http://[::1]:9411/greet" -> Docker (the lo0 interface on ipv6)
```
But if we try the `en0` interface on ipv6 then we get Helidon:

```
 curl "http://[2601:582:c200:f0b7:180c:fa42:289b:eaf2]:9411/greet"
{"message":"Hello World!"}
```

It follows that the Helidon app succeeded because it found one interface/protocol combo available to bind. We can make the Helidon app fail forcing the use of tcp4 using the following property:

```
-Djava.net.preferIPv4Stack=false
```

given that this reduces the search space for a possible binding. 

To summarize, this seems to be working as (somebody) designed, it is just an odd combination of bindings defined by Docker that results in this problem.

**Comment by spericas** [comment]
After further research, I think there's not much we can do to alter this behavior (other than using the property shown above). My recommendation is that we put this in a README or FAQ.
